File: java-reactive-logger/src/test/java/hu/numichi/reactive/logger/java/MDCContextTest.java
package hu.numichi.reactive.logger.java;

import hu.numichi.reactive.logger.MDC;
import hu.numichi.reactive.logger.exception.InvalidContextDataException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;
import reactor.util.context.Context;

import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

import static hu.numichi.reactive.logger.Consts.DEFAULT_REACTOR_CONTEXT_MDC_KEY;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MDCContextTest {
    
    public static final String ANOTHER_CONTEXT_KEY = "another-context-key";
    
    @Test
    @DisplayName("should give the MDC you are looking for")
    void shouldGiveTheMDCByDefault() {
        MDC mdc1 = new MDC();
        mdc1.put("mdcKey", "mdcValue");
    
        MDC mdc2 = new MDC(ANOTHER_CONTEXT_KEY);
        mdc2.put("mdcKey", "mdcValue");
    
    
        Mono<MDC> resultDefaultByContext = Mono.deferContextual(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
    
        StepVerifier.create(resultDefaultByContext)
            .expectNext(mdc1)
            .verifyComplete();
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
        
        StepVerifier.create(resultDefault)
            .expectNext(mdc1)
            .verifyComplete();
    
    
        Mono<MDC> resultAnotherByContext = Mono.deferContextual(ctx -> MDCContext.read(ctx, ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
    
        StepVerifier.create(resultAnotherByContext)
            .expectNext(mdc2)
            .verifyComplete();
        
    
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
    
        StepVerifier.create(resultAnother)
            .expectNext(mdc2)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should give the MDC you are looking for (more MDC Context)")
    void shouldGiveTheMDCWithMoreMdcContext() {
        MDC defaultMdc = new MDC();
        defaultMdc.put("mdcKey", "mdcValue");
    
        MDC anotherMdc = new MDC(ANOTHER_CONTEXT_KEY);
        anotherMdc.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
    
        StepVerifier.create(resultDefault)
            .expectNext(defaultMdc)
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        
        StepVerifier.create(resultAnother)
            .expectNext(anotherMdc)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should need to be able to store a map")
    void shouldNeedToBeAbleToStoreAMap() {
        Map<String, String> mdcMap = new HashMap<>();
        mdcMap.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdcMap));
    
        StepVerifier.create(resultDefault)
            .expectNextMatches(mdc1 ->  mdc1.getMap().equals(mdcMap) && mdc1.getContextKey().equals(DEFAULT_REACTOR_CONTEXT_MDC_KEY))
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, ANOTHER_CONTEXT_KEY, mdcMap));
        
        StepVerifier.create(resultAnother)
            .expectNextMatches(mdc1 ->  mdc1.getMap().equals(mdcMap) && mdc1.getContextKey().equals(ANOTHER_CONTEXT_KEY))
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should be MDC stored with the overridden context ID")
    void shouldBeStoredWithTheOverriddenContextIDTest() {
        MDC mdc = new MDC();
        mdc.put("mdcKey", "mdcValue");
        
        Mono<MDC> result = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, ANOTHER_CONTEXT_KEY, mdc));
        
        StepVerifier.create(result)
            .expectNextMatches(mdc1 ->  mdc1.getMap().equals(mdc.getMap()) && mdc1.getContextKey().equals(ANOTHER_CONTEXT_KEY))
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should throw NoSuchElementException if context key is not exist")
    void shouldThrowNoSuchElementExceptionTest() {
        MDC mdc = new MDC(ANOTHER_CONTEXT_KEY);
        mdc.put("mdcKey2", "mdcValue2");
        
        
        Mono<MDC> result1 = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc));
    
        StepVerifier.create(result1).expectError(NoSuchElementException.class).verify();
        
        
        Mono<MDC> result2 = Mono.defer(() -> MDCContext.read("not-exist-context-id"))
            .contextWrite(it -> MDCContext.put(it, mdc));
        
        StepVerifier.create(result2).expectError(NoSuchElementException.class).verify();
    }
    
    @Test
    @DisplayName("should throw NullPointerException if any parameter is NULL")
    void shouldThrowNullPointerExceptionTest() {
        assertThrows(NullPointerException.class, () -> MDCContext.put(null, "", new HashMap<>()));
        assertThrows(NullPointerException.class, () -> MDCContext.put(Context.of("", ""), null, new HashMap<>()));
        assertThrows(NullPointerException.class, () -> MDCContext.put(Context.of("", ""), "", null));
        
        StepVerifier.create(MDCContext.read((String) null))
            .expectError(NullPointerException.class)
            .verify();
    
        StepVerifier.create(MDCContext.read(null, ""))
            .expectError(NullPointerException.class)
            .verify();
        
        StepVerifier.create(MDCContext.read(Context.of("", ""), null))
            .expectError(NullPointerException.class)
            .verify();
    }
    
    @Test
    @DisplayName("should throw InvalidContextDataException if context data is invalid")
    void shouldThrowInvalidContextDataException() {
        Mono<MDC> result2 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, ""));
    
        Mono<MDC> result3 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, 100));
    
        StepVerifier.create(result2)
            .expectError(InvalidContextDataException.class)
            .verify();
    
        StepVerifier.create(result3)
            .expectError(InvalidContextDataException.class)
            .verify();
    }
}

File: kotlin-coroutine-reactive-logger/src/test/kotlin/hu/numichi/reactive/logger/kotlin/MDCContextTest.kt
package hu.numichi.reactive.logger.kotlin

import hu.numichi.reactive.logger.Consts
import hu.numichi.reactive.logger.Consts.DEFAULT_REACTOR_CONTEXT_MDC_KEY
import hu.numichi.reactive.logger.MDC
import hu.numichi.reactive.logger.exception.InvalidContextDataException
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.reactor.ReactorContext
import kotlinx.coroutines.reactor.asCoroutineContext
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import reactor.util.context.Context

@ExperimentalCoroutinesApi
internal class MDCContextTest {

    companion object {
        const val ANOTHER_CONTEXT_KEY = "another-context-key"
    }

    @Test
    fun `should give the MDC you are looking for (default key)`() = runTest {
        val mdc = MDC()
        mdc["mdcKey"] = "mdcValue"

        MDCContext.withContext(mdc) {
            val result = MDCContext.read()
            assertEquals(mdc, result)
        }

        MDCContext.withContext(mdc) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context)
            assertEquals(mdc, result)
        }

        MDCContext.withContext(Context.of(mapOf("1" to "2")), mdc) {
            val size = coroutineContext[ReactorContext]?.context?.size()
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context)
            assertEquals(mdc, result)
            assertEquals(2, size)
        }
    }

    @Test
    fun `should give the MDC you are looking for (default key + use native withContext)`() = runTest {
        val mdcMap = mutableMapOf<String, String>()
        mdcMap["mdcKey"] = "mdcValue"

        withContext(Context.of(mapOf(DEFAULT_REACTOR_CONTEXT_MDC_KEY to mdcMap)).asCoroutineContext()) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context)
            assertEquals(mdcMap, result)
            assertEquals(mdcMap, result.map)
        }
    }

    @Test
    fun `should give the MDC you are looking for (another key)`() = runTest {
        val mdc2 = MDC(ANOTHER_CONTEXT_KEY)
        mdc2["mdcKey"] = "mdcValue"

        MDCContext.withContext(mdc2) {
            val result = MDCContext.read(ANOTHER_CONTEXT_KEY)
            assertEquals(mdc2, result)
        }
    }

    @Test
    fun `should give the MDC you are looking for (more MDC Context)`() = runTest {
        val mdc1 = MDC()
        mdc1["mdcKey"] = "mdcValue"

        val mdc2 = MDC(ANOTHER_CONTEXT_KEY)
        mdc2["mdcKey"] = "mdcValue"

        MDCContext.withContext(mdc1, mdc2) {
            val result1 = MDCContext.read(coroutineContext[ReactorContext]?.context)
            val result2 = MDCContext.read(coroutineContext[ReactorContext]?.context, DEFAULT_REACTOR_CONTEXT_MDC_KEY)
            val result3 = MDCContext.read(DEFAULT_REACTOR_CONTEXT_MDC_KEY)
            val result4 = MDCContext.read(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY)
            val result5 = MDCContext.read(ANOTHER_CONTEXT_KEY)

            assertEquals(mdc1, result1)
            assertEquals(mdc1, result2)
            assertEquals(mdc1, result3)
            assertEquals(mdc2, result4)
            assertEquals(mdc2, result5)
        }
    }

    @Test
    fun `should need to be able to store a map`() = runTest {
        val mdcMap = mutableMapOf<String, String>()
        mdcMap["mdcKey"] = "mdcValue"

        MDCContext.withContext(MDC(mdcMap)) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context).map
            assertEquals(mdcMap, result)
        }

        MDCContext.withContext(MDC(ANOTHER_CONTEXT_KEY, mdcMap)) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY).map
            assertEquals(mdcMap, result)
        }
    }

    @Test
    fun `should be MDC stored with the overridden context ID`() = runTest {
        val mdcMap = mutableMapOf<String, String>()
        mdcMap["mdcKey"] = "mdcValue"

        MDCContext.withContext(MDC(mdcMap)) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context).map
            assertEquals(mdcMap, result)
        }

        MDCContext.withContext(MDC(ANOTHER_CONTEXT_KEY, mdcMap)) {
            val result = MDCContext.read(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY).map
            assertEquals(mdcMap, result)
        }
    }

    @Test
    fun `should throw NullPointerException if any parameter is NULL`() = runTest {
        assertThrows<NullPointerException> {
            MDCContext.read(null)
        }

        assertThrows<NullPointerException> {
            MDCContext.read(null, ANOTHER_CONTEXT_KEY)
        }
    }

    @Test
    fun `should throw InvalidContextDataException if context data is invalid`() = runTest {
        withContext(Context.of(mapOf(Consts.DEFAULT_REACTOR_CONTEXT_MDC_KEY to 10)).asCoroutineContext()) {
            assertThrows<InvalidContextDataException> { MDCContext.read() }
        }

        withContext(Context.of(mapOf(Consts.DEFAULT_REACTOR_CONTEXT_MDC_KEY to "any")).asCoroutineContext()) {
            assertThrows<InvalidContextDataException> { MDCContext.read() }
        }
    }
}